uber/site_sections/registration.py

```diff
@@ -863,6 +864,37 @@ def feed(self, session, page='1', who='', what='', action=''):
             'who_opts': [who for [who] in session.query(Tracking).distinct().order_by(Tracking.who).values(Tracking.who)]
         }
 
+    @csrf_protected
+    def undo_delete(self, session, id, message='', page='1', who='', what='', action=''):
+        if cherrypy.request.method == "POST":
+            model_class = None
+            tracked_delete = session.query(Tracking).get(id)
+            if tracked_delete.action != c.DELETED:
+                message = 'Only a delete can be undone'
+            else:
+                model_class = getattr(sa, tracked_delete.model)
+
+            if model_class:
+                params = json.loads(tracked_delete.snapshot)
+                model_id = params.get('id').strip()
+                if model_id:
+                    existing_model = session.query(model_class).filter(
+                        model_class.id == model_id).first()
+                    if existing_model:
+                        message = '{} has already been undeleted'.format(tracked_delete.which)
+                    else:
+                        model = model_class(id=model_id).apply(params, restricted=False)
+                else:
+                    model = model_class().apply(params, restricted=False)
+
+                if not message:
```

Comment:
I would recommend adding an extra safety check here (or during apply...? Not sure where it will fail) just in case there's some deep problem with the data saving to the database. We want the session rolled back in that case and an error message displayed (something like: "Error undeleting this {model}. Did something about the database change since it was deleted?"). This would be just in case we:

- remove a column
- add a non-nullable column
- change a set of possible Choice/MultiChoice options (this is more common than you'd think)

between when the model was deleted and when someone tries to undelete it.